### ISSUE TITLE
Card Search Issues on OS Ubuntu 14.04 LTS Linux

### ISSUE THREAD

**Issue by Passione232** [Cockatrice/Cockatrice#1720]
When I try to look up cards, the window freezes for a bit when I enter the first few letters, then returns to normal. When I erase it does the same thing as well, freezing before finally erasing the last few letters.

**Comment by ctrlaltca** [comment]
It's actually filtering a long list of items, so it can take some time. I coded a "progressive loading" patch some time ago, i can try to open a PR with it and see if it fixes the problem for you.

**Comment by tooomm** [comment]
Which hardware do you use?

**Comment by Daenyth** [comment]
What kind of CPU and RAM do you have? It's doing a fair amount of work but it shouldn't be *that* slow.

**Comment by Passione232** [comment]
Intel(R) Celeron(R) CPU N2840 @ 2.16GHz
Total Memory	3931056 kB

**Comment by Passione232** [comment]
And I would like the Progressive Loading Patch you mentioned Ctrl.

**Comment by Daenyth** [comment]
Fairly low end but it shouldn't be so slow even on that. There's definitely some weirdness.

**Comment by ctrlaltca** [comment]
The progressive patch is in #1721.

**Comment by Passione232** [comment]
So what do I do now?

**Comment by Daenyth** [comment]
@Passione232 did you compile yourself before or use the ubuntu deb? You can compile from the pr branch by checking it out with git.

```
git clone https://github.com/ctrlaltca/Cockatrice.git cockatrice-speedup
cd cockatrice-speedup
git checkout origin/speedup
mkdir build
cd build
```

And then from there do cmake, make, etc as normal for ubuntu. Detailed steps on that here: https://github.com/Cockatrice/Cockatrice/wiki/Compiling-Cockatrice#ubuntu-14xx

Just skip the step that includes `git clone` and go right to the step with `cmake`. You may still need to run the step before that which `apt-get install`s the dependencies


@ctrlaltca is there a way we can manually trigger travis/bintray to build a deb for a given pr branch?

**Comment by ctrlaltca** [comment]
@Daenyth the deb gets already built, but it's not upload anywhere.

**Comment by Daenyth** [comment]
Can it be set up to be uploaded to bintray in a way where it's only available for public download if one of us manually blesses it?

Or maybe if we make a branch on the official repo following yours and push, then would that get uploaded to bintray since it's in the official repo?

My goal here is to streamline the ability for us to offer downloadable installers for unmerged features so that we don't have to build and upload ourselves.

**Comment by ctrlaltca** [comment]
Yep, i understand the point. This can already work for windows packages (appveyor stores the artifacts for some months), i'll see what i can do for osx/linux packages.

**Comment by Daenyth** [comment]
+ :100:

**Comment by Passione232** [comment]
@Daenyth I did as you told me to, however the problem persists. When I was running Cockatrice on Windows, there was no delay of this kind, so it can't be my laptop. What do you suggest I do?

**Comment by Daenyth** [comment]
Interesting. Were you using qt4 or qt5 on windows? xp/vista version or 7+?

**Comment by Passione232** [comment]
Windows 8 and then Windows 10. I am not sure what Qt I was using.

**Comment by Daenyth** [comment]
I think that would be qt5 then. Can you post the output of cmake from
compiling with the instructions?

**Comment by Passione232** [comment]
terminalzero@VizkopaTerminal:~/cockatrice-speedup/build$ cmake .. -DWITH_SERVER=1
-- UPDATE TRANSLATIONS: OFF
-- Found Qt 5.2.1
-- Mysql connector NOT FOUND: servatrice won't be able to connect to a mysql server
-- Configuring done
-- Generating done
-- Build files have been written to: /home/terminalzero/cockatrice-speedup/build
Output of CMake

**Comment by ZeldaZach** [comment]
Looks to me like you don't have mysql installed. Try installing `libqt5sql5-mysql`

**Comment by Daenyth** [comment]
@ZeldaZach mysql is only needed for servatrice. It has no impact on this.

**Comment by ZeldaZach** [comment]
Oh, I thought he was trying to install servatrice. Sorry

**Comment by Passione232** [comment]
@Daenyth Where have we gotten to with the problem [I.E] solving it?

**Comment by Daenyth** [comment]
Nothing at the moment.

**Comment by Passione232** [comment]
@Daenyth Might be a bit of a necrobump but I'd really like if something could be done about this issue. I really like using Cockatrice, and frankly, XMage just feels so slow and clunky compared to it. Not to mention a new update which pretty much threw the resolution out of whack, but I digress.

**Comment by Daenyth** [comment]
@Passione232 We've just released a new version, could you try it and see if it's still a problem?

https://bintray.com/artifact/download/cockatrice/Cockatrice/Cockatrice-Gatewatchers-trusty-ubuntu.deb

No one here has been able to reproduce your issue, I believe.

**Comment by Passione232** [comment]
Yes, apparently the delay still exists. Would it be helpful if I recorded the problem through a screencast video?

**Comment by Daenyth** [comment]
It won't hurt. Can you try the windows version in wine for now?

**Comment by ziobbe** [comment]
I'm not sure if i should open another issue or just comment here :p
I think i have the same problem: i installed cockarice  two days ago (after a format) from the .deb and when i search a card the program freeze and remain unusable for several seconds for every letter i insert or delete, on my previous installation this didn't happen, what can i do?

**Comment by tooomm** [comment]
@ziobbe What's your exact OS? Which version of Cockatrice are you running?

**Comment by ziobbe** [comment]
Right now i'm running elementary os freya, wich is based on ubuntu 14.04, Cockatrice version is 277d7e2 (2016-06-30)

**Comment by tooomm** [comment]
Are you possibly on some older hardware?

**Comment by ziobbe** [comment]
It's an i3 dell inspiron, i played some WoW and other games or emulators no problem.
I have used cockatrice on this machine in the past without running in this kind of issues.

**Comment by tooomm** [comment]
Do you possibly know which version you used before? Was it a main release? When did you download/install it?

**Comment by ziobbe** [comment]
I found a build directory on my backup containing Cockatrice-0.0.1~git20151230.924bbb2.deb i think this is the version i used previously, the metadata says it was last modified January 3.

**Comment by Daenyth** [comment]
This probably won't fix it, but if you have it installed try `apt-get remove appmenu-qt5`

Was one of the dependencies qt4? I'm wondering if this is from when we switched the default to qt5.

**Comment by Daenyth** [comment]
Actually it also looks like you have 4gb ram - what size is your card picture cache set to? I wonder if the slowness is caused by memory exhaustion forcing you to swap to disk. Try setting your card picture cache small, down to something like 256mb. I think by default it might be set to 4gb

**Comment by ziobbe** [comment]
I tried the command and received the output "Package 'appmenu-qt5' is not installed, so not removed
".
The missing package for the installer of the previous version is libprotobuf9v5.
Changing the image chace to 256 mb (original was 2047mb)  doesn't seem to make any improvements, I even tried to disable the image download function without noticable differences in performance.
I've been monitoring with htop the cpu usage and when I filter the database writing in the searchbox on the first letter the cpu spikes to 100% usage while the application is unresponsive for several seconds.
Thanks for your help and attention =).

**Comment by Daenyth** [comment]
The CPU spike is interesting.

Can you paste the entire command line output from running the app manually with `cockatrice` in your shell?

Another thing we can try, can you use the "Edit Sets" feature in the deck editor to disable everything except 1-2 sets and see if that changes the speed?

**Comment by ziobbe** [comment]
Hi, the first letter is the most noticeble, the most popular letters cause more lag (x had low lag), deleting till the first letter brings the same lag as inserting the first letter and deleting all the letters has the maximum effect (i noticed this deleting x), using the filter option is slow in the same way.
In theese days i diabled all the sets i don't use and this seems to mitigate the problem, using a few set only for example reduce the problem to zero, disabling eldritch moon and shadow over innistrad doesn't remove the problem.
The most noticable thing in htop is that cockatrice appears several times and when the spikes occur it happens on one core only, the memory impact isn't stunning.
Launching cockatrice from the command line produces no output at all and the debug log doesnt' say anything interesting.

**Comment by Daenyth** [comment]
No output at all? That's weird. It should at least mention things like launching ui and such.

The description of performance behavior makes me pretty sure this is because of #2054, so I think at this point we at least know where we need to make changes to improve it for you.

I don't know exactly when we will be able to implement the changes needed.

For now, a workaround for you might be to use the advanced filters from the middle panel. In a nutshell, the fast search is filtering the entire card list on every single letter change whereas the advanced filters are applied all at once.

**Comment by Daenyth** [comment]
Also showing up multiple times is because it uses threads, I believe it has one thread for the UI, another for talking to the server, and another for downloading pictures

**Comment by tooomm** [comment]
>when the spikes occur it happens on one core only

can we parallelize this, like different cores/threads search in different (enabled) sets or parts of the card list?

**Comment by Daenyth** [comment]
I don't think that would help much, the threads would have to talk to each other a lot

**Comment by ziobbe** [comment]
I'm not sure to have understood the workaround you propose, strangely enaugh applying filters is as much slow as namesearching (with the cpu spiking to 100% on one core for some second and the application unfreezing seconds after the spike is disappeared).

**Comment by Daenyth** [comment]
I guess that makes sense; I think the problem is that the filtering (both fast and advanced filters) start with the entire card database and have to check every entry.

**Comment by ziobbe** [comment]
The thing i do not understand is that on my tower wich has worse hardware cockatrice works like a charm, and even on this dell it worked as intended before i formatted it, anyway thank you all so much for your time and effort, i really like this community =).
I hope to be able to use cockatrice on the go and at uni while communting sooner or later.

**Comment by Daenyth** [comment]
As far as I'm aware you're the only person who's seen this issue. It's hard to know why it affects you so drastically more than everyone else.

**Comment by ziobbe** [comment]
I see, thanks for the help, is there a way to know when there is an updat on the matter or a general update to know if anything changed?

**Comment by ctrlaltca** [comment]
> is there a way to know when there is an updat on the matter or a general update to know if anything changed?

We'll update this ticket, so you should receive an email notification.

**Comment by tooomm** [comment]
Any update here? Does this problem persists with recent versions?

**Comment by ebbit1q** [comment]
This is an issue with how the cards are displayed, and it hasn't changed one bit. A solution would be to simply display the first n items and fetch more when scrolling down.

**Comment by BruebachL** [comment]
Assumed fixed by #5974
